### ISSUE TITLE
NuGet Version Number Out of Sync with Assembly Version

### ISSUE THREAD

**Issue by spottedmahn** [dotnet/corefx#29349]
Why are the NuGet versions and the assembly versions not in sync?

### Example

Assembly | NuGet Version | Assembly Version
------ | --------- | --------
[System.Net.Http](https://www.nuget.org/packages/System.Net.Http/) | 4.3.3 | 4.1.1.2

Is this across the board or just some assemblies?

---

I have to go through mental gymanistics when I get exceptions like this:

>System.IO.FileLoadException: Could not load file or assembly 'System.Net.Http, Version=4.1.1.2, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. 

- Why NuGet package do I have installed?
- Ok, that corresponds to which assembly version?  Crack open IL Spy...
...

**Comment by ericstj** [comment]
Package versions correspond to releases.  Assembly versions correspond to APIs/bugfixes/servicing.  We used to version the packages to match the (maximum) assembly version and we had a different sort of problem: which set of package versions should I reference to get all of the 1.1 release?  Different assembly versions exist in a single package because the package must not drop support for all frameworks and some frameworks have a specific version inbox, whereas other frameworks must have a unique version for every release so that the assemblies don't fight over a single slot in the GAC.  Since we already had this disconnect of mapping many versions to one folks said it doesn't make much difference to completely divorce package version and assembly version.

Adding the assembly information to the NuGet spec is one thing we could do, another thing might be to put the package version in the assembly metadata.  Both seem like reasonable suggestions to put on the backlog.

**Comment by spottedmahn** [comment]
What is the next step here?

**Comment by ericstj** [comment]
Let us know if your question was answered and if you think the suggestions I made would improve the situation.

**Comment by spottedmahn** [comment]
I love 💖 the suggestions!  Do I need to create any more GitHub issues for these suggestions?

> - Adding the assembly information to the NuGet spec is one thing we could do
> - Another thing might be to put the package version in the assembly metadata

How I do follow the progress?

Should I close this Issue?

**Comment by ericstj** [comment]
Ok, for now let's use this issue to track it.  Eventually the work will need to happen in buildtools.  I discussed with @joperezr and he mentioned he can pick it up.

**Comment by spottedmahn** [comment]
Cool 😎, works for me.  Thanks for timely responses!

**Comment by ViktorHofer** [comment]
> Another thing might be to put the package version in the assembly metadata

That isn't possible as we are harvesting many assemblies. If we put the package version in the assembly's metadata it would point to an old package version. I think we shouldn't do this.

> Adding the assembly information to the NuGet spec is one thing we could do

Which assembly version would we use? We have many different tfms in the package that contain different assembly versions (because of harvesting). Just always the latest?

**Comment by ericstj** [comment]
Assembly metadata: sure but first package it shipped it could be good enough.  Depends on how folks expect to to use it.

Package spec: We'd have a table similar to the package report where it can list the version applicable per framework.

**Comment by ViktorHofer** [comment]
@ericstj should we close this in favor of https://github.com/dotnet/corefx/issues/40544? That won't cover all OOB packages but probably the majority.

**Comment by joperezr** [comment]
Not really. The problem with this issue is that OOB packages contain many assemblies in them, for different frameworks, and in most cases assembly versions would vary between the different assemblies in the package. With the issue you are mentioning, all that it means is that one of the assemblies might match the package version, but you still lack the info on all assembly version contained within a package.
